packages/lodestar/src/chain/validation/contributionAndProof.ts

```diff
@@ -0,0 +1,105 @@
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {altair} from "@chainsafe/lodestar-types";
+import {IBeaconDb} from "../../db";
+import {IContributionAndProofJob, SyncCommitteeError, SyncCommitteeErrorCode} from "../errors/syncCommitteeError";
+import {IBeaconChain} from "../interface";
+
+export async function validateGossipContributionAndProof(
+  config: IBeaconConfig,
+  chain: IBeaconChain,
+  db: IBeaconDb,
+  contributionAndProofJob: IContributionAndProofJob
+): Promise<void> {
+  const signedContributionAndProof = contributionAndProofJob.contributionAndProof;
+  const contributionAndProof = signedContributionAndProof.message;
+  const contribution = contributionAndProof.contribution;
+  // The contribution's slot is for the current slot
+  if (chain.clock.currentSlot < contribution.slot) {
+    throw new SyncCommitteeError({
+      code: SyncCommitteeErrorCode.FUTURE_SLOT,
+      currentSlot: chain.clock.currentSlot,
+      syncCommitteeSlot: contribution.slot,
+      job: contributionAndProofJob,
+    });
+  }
+
+  if (chain.clock.currentSlot > contribution.slot) {
+    throw new SyncCommitteeError({
+      code: SyncCommitteeErrorCode.PAST_SLOT,
+      currentSlot: chain.clock.currentSlot,
+      syncCommitteeSlot: contribution.slot,
+      job: contributionAndProofJob,
+    });
+  }
+
+  // The block being signed over has been seen
+  if (!chain.forkChoice.hasBlock(contribution.beaconBlockRoot)) {
+    throw new SyncCommitteeError({
+      code: SyncCommitteeErrorCode.UNKNOWN_BEACON_BLOCK_ROOT,
+      beaconBlockRoot: contribution.beaconBlockRoot as Uint8Array,
+      job: contributionAndProofJob,
+    });
+  }
+
+  if (contribution.subCommitteeIndex >= altair.SYNC_COMMITTEE_SUBNET_COUNT) {
+    throw new SyncCommitteeError({
+      code: SyncCommitteeErrorCode.INVALID_SUB_COMMITTEE_INDEX,
+      subCommitteeIndex: contribution.subCommitteeIndex,
+      job: contributionAndProofJob,
+    });
+  }
+
+  if (db.seenSyncCommiteeCache.hasContributionAndProof(contributionAndProof)) {
+    throw new SyncCommitteeError({
+      code: SyncCommitteeErrorCode.SYNC_COMMITTEE_ALREADY_KNOWN,
+      root: config.types.altair.ContributionAndProof.hashTreeRoot(contributionAndProof),
+      job: contributionAndProofJob,
+    });
+  }
+  // check the validator is aggregator for the slot
+  if (!isSyncCommitteeAggregator(config, contributionAndProof.selectionProof)) {
+    throw new SyncCommitteeError({
+      code: SyncCommitteeErrorCode.INVALID_AGGREGATOR,
+      aggregatorIndex: contributionAndProof.aggregatorIndex,
+      job: contributionAndProofJob,
+    });
+  }
+
+  const headState = chain.getHeadState();
+  const aggregatorPubkey = headState.validators[contributionAndProof.aggregatorIndex].pubkey;
+  if (
+    !getSyncSubCommitteePubkeys(
+      config,
+      headState as CachedBeaconState<altair.BeaconState>,
+      contribution.subCommitteeIndex
+    ).includes(aggregatorPubkey)
```

Comment:
Is this approach efficient enough?
- What's the cost of the getSyncSubCommitteePubkeys() utility? We may have to run it 1000 times per slot if we have enough validators
- Note that doing an .includes requires checking the entire array and is O(n)

Reply:
+ `getSyncSubCommitteePubkeys` utility just extract pubkeys from a bigger `pubkeys` array inside `state.currentSyncCommittee` so it's not expensive.
But we should cache it as it's only changed per ~27h, I believe we'll create a cache for it in `CachedBeaconState`. I'd mark a TODO for this and wait for @wemeetagain fast state transition work for altair
+ once we have the cache, we just do `subCommitteeIndices.includes(aggregatorIndex)` so `.includes` is not a concern here